### PR TITLE
[ticket/16879] Add events to phpbb_mail()

### PR DESCRIPTION
**Pull request by 3D-I**
PHPBB3-16879

Checklist:

- [x] Correct branch: master for new features; 3.3.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/master/coding-guidelines.html) and [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

https://tracker.phpbb.com/browse/PHPBB3-16879

### CODE REVIEW

**Comment by marc1706** [comment]
Events will not be added between RC and stable versions for a patch level release.

File: phpBB/includes/functions_messenger.php [hunk 1]
```diff
@@ -1935,8 +1935,54 @@ function phpbb_mail($to, $subject, $msg, $headers, $eol, &$err_msg)
 	 */
 	$additional_parameters = $config['email_force_sender'] ? '-f' . $config['board_email'] : '';
 
+	/**
+	 * Modify data before sending out emails with the PHP's mail function
```
Comment (marc1706) [inline phpBB/includes/functions_messenger.php:1939]:
```suggestion
	 * Modify data before sending out emails with PHP's mail function
```

File: phpBB/includes/functions_messenger.php [hunk 2]
```diff
@@ -1935,8 +1935,54 @@ function phpbb_mail($to, $subject, $msg, $headers, $eol, &$err_msg)
...
+		'headers',
+		'eol',
+		'additional_parameters',
+	];
+	extract($phpbb_dispatcher->trigger_event('core.phpbb_mail_before', compact($vars)));
+
 	$result = mail($to, mail_encode($subject, ''), wordwrap(utf8_wordwrap($msg), 997, "\n", true), $headers, $additional_parameters);
 
+	/**
+	 * Execute code after sending out emails with the PHP's mail function
```
Comment (marc1706) [inline phpBB/includes/functions_messenger.php:1963]:
```suggestion
	 * Execute code after sending out emails with PHP's mail function
```

**Review by CHItA — APPROVED** [review]

**Comment by rubencm** [comment]
Why is needed one for before and another for after?
Only before event dont cover all use cases?
@3D-I 

Also wouldn't be better to just dispatch one event before sending email (independently if is with smtp or mail)?

**Comment by 3D-I** [comment]
@rubencm The `after` is needed because of the `result`. If false (mail not sent) we can use the same event to try and send the email using another transport or store it locally (for testing purposes, as an example). Useful if testing extensions locally, as I do.

**Pull request merged**